### PR TITLE
trees: Separate out leaves - prep for proba output and soft voting

### LINKED ISSUE #12: RFC: Inconsistent results when using `max_depth` - support soft voting
As mentioned in #9 , when limiting the depth of a RFC using `max_depth`, there are inconsistencies in the predictions. I've discussed the same issue here: https://github.com/nok/sklearn-porter/issues/52 

I've dived a bit into the issue and figured that it is indeed what I expected: The final leave probabilities are only 0 and 1 if the tree if fully grown. Therefore I thought I make a separate issue here to leave the information somewhere.

https://stackoverflow.com/questions/30814231/using-the-predict-proba-function-of-randomforestclassifier-in-the-safe-and-rig
 (see comments)

> 1) About prediction precision: I insist but this is not a question of number of trees. Even with a single decision tree you should be able to get probability predicitions with more than one digits. A decision tree aims at clustering the inputs based on some rules (the decision), and these clusters are the leafs of the tree. If you have a leaf with 2 non-spam emails and one spam email from your training data, then the probability prediction for any email that belongs to this leaf/cluster (with regards to the rules established by fitting the model), is : 1/3 for spam and 2/3 for non-spam. – Sebastien Jun 20 '15 at 14:49
>2) About the dependencies in predictions: Again Sklearn definition gives the answer : the probability is computed with regards to the leaf (corresponding to your email to test) 's characteristics : the number of instances of each class in this leaf. This is set when your model is fitted, so it only depends on the training data. In conclusion : the result is the probability of instance 1 to spam with 60% whatever the other 9 instances' probabilities are. – Sebastien Jun 20 '15 at 15:00

and 

https://stats.stackexchange.com/questions/193424/is-decision-tree-output-a-prediction-or-class-probabilities

>So depending on implementation: predicted probability is either (a) the mean terminal leaf probability across all trees or (b) the fraction of trees voting either class. If out-of-bag(OOB) prediction, then only in trees where sample is OOB. For a single fully grown tree, I would guess the predicted probability only could be 0 or 1 for any class, because all terminal nodes are pure(same label). If the single tree is not fully grown and/or more trees are grown, then the predicted probability can be a positive rational number from 0 to 1.

As I assume that it is quite a bit of re-work to integrate this changes without losing too much performance (e.g. returning pointers to arrays instead of class indexes by the individual trees), I don't want to force you to fix it just for me or so ;-) I just wanted to mention it here. So it's fine with me if it remains like it is for now.

### PR DESCRIPTION
**Pull request by jonnor**
This moves the leaves out of the nodes and into a separate array, such that they can have different sizes depending on the desired output. For soft-voting one needs to store class probabilities. Preparation for fixing #12 

DONE
* [x] Move leaves out of tree
* [x] Add test for soft-voting case. 
* [x] Add tests for predict_proba()

TODO
* [ ] Benchmark to check that performance did not regress (inference time and model size)
* [ ] Make soft-voting the default behavior (for maximum compatibility)
* [ ] Add documentation on leaf_bits / hard voting.
* [ ] And example of using leaf_bits

### CODE REVIEW

**Comment by jonnor** [comment]
Many changes across the code related to decision tree ensembles here. Will merge the leaves split as-is, and then finalizing soft voting / proba output can happen in a follow-up MR.
